docs/strings.md

```diff
@@ -0,0 +1,464 @@
+   For example, should Japanese text in Big5 encoding be required to convert to Unicode for all operations that are not part of the class's internal implementation?
+ - Should strings that are using a Unicode encoding expose interfaces in a *specific* unicode encoding?
+
+*Proposal*: Standard library strings should expose interfaces in 32-bit Unicode code points.
+They should also be able to export or provide views in their native encoding so that operations can be optimised based on pattern matching.
+
+Draft proposal for a core string library
+----------------------------------------
+
+The Verona standard string should define a string interface and a small set of concrete instantiations of this interface.
+The core string class should implement the minimum required for generic implementation of string operations, which can be implemented either using dynamic dispatch over the string interface or as generic reifications over the concrete types.
+We should make it possible to support a rich Unicode library that uses the same string objects as the standard library (and third-party libraries that use the same interface) and provides functions for iteration with unicode graphemes, locale-aware collation and other more complex operations that not all programs need.
+
+The interface for strings would look roughly like this:
+
+```verona
+/**
+ * String contains only 7-bit ASCII characters and is stored with each code
+ * point in an 8-bit integer.
+ */
+class StringEncodingASCII
+{
+	typedef CharType = U8;
+}
+/**
+ * String is stored as UTF-8-encoded text.
+ */
+class StringEncodingUTF8
+{
+	typedef CharType = U8;
+}
+/**
+ * String is stored as UTF-16-encoded text.
+ */
+class StringEncodingUTF16
+{
+	typedef CharType = U16;
+}
+/**
+ * String is stored as UTF-32-encoded text.
+ */
+class StringEncodingUTF32
+{
+	typedef CharType = U32;
+}
+/**
+ * Interface for other string encodings.  Any other type can be used as a valid
+ * string encoding but code that does not have any special cases for that
+ * encoding will be able to treat is as unknown.
+ */
+interface StringEncodingCustom
+{
+	typedef CharType;
+}
+
+/**
+ * The set of Unicode encodings.  All standard library functions that have
+ * special cases for specific encodings should handle these.
+ */
+typedef UnicodeStringEncoding = StringEncodingASCII |
+                                StringEncodingUTF8 |
+                                StringEncodingUTF16 |
+                                StringEncodingUTF32;
+/**
+ * The full set of possible string encodings, can be extended by third-party code.
+ */
+typedef StringEncoding = StringEncodingASCII |
+                         StringEncodingUTF8 |
+                         StringEncodingUTF16 |
+                         StringEncodingUTF32 |
+                         StringEncodingCustom;
+
+/**
+ * Character type, stores a unicode code point.
+ */
+typedef Unichar = U32;
```

Comment:
Pure bikeshedding, so please ignore: I actually like the Go use of "Rune", so maybe something like `type Rune = U32;`.

Reply:
POSIX also spells it Rune, so that might make sense here too, but I always think of [Hugo Rune](https://en.wikipedia.org/wiki/Hugo_Rune) when I read 'rune'.